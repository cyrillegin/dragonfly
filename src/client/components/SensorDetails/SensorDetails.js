import React, { useState } from 'react';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import { AddAction } from '../Modals';
import { actionConditions, actionProperties } from '../../utilities/constants';
import { windowEmitter } from '../../utilities/Window';
import Api from '../../Api';

const SensorDetails = ({ className, sensor }) => {
  const [sensorDetails, setSensorDetails] = useState({
    name: sensor.name || '',
    description: sensor.description || '',
    coefficients: sensor.coefficients || '',
    pollRate: sensor.pollRate || '',
    unit: sensor.unit || '',
  });

  const [successMessage, updateSuccessMessage] = useState('');
  const [actionModal, updateActionModal] = useState({});
  const [actionMessage, setActionMessage] = useState('');

  const addAction = action => {
    updateActionModal({ id: -1 });
  };

  const editAction = action => {
    updateActionModal(action);
  };

  const deleteAction = action => {
    Api.deleteAction(action.id).then(() => {
      windowEmitter.emit('station-refresh');
    });
  };

  const saveAction = action => {
    const body = {
      stationId: sensor.stationId,
      sensorId: sensor.id,
      ...action,
    };
    let promise;
    if (action.id === -1) {
      delete body.id;
      promise = Api.createAction(body);
    } else {
      promise = Api.updateAction(body);
    }

    promise.then(res => {
      if (res.error) {
        setActionMessage(res.error);
      } else {
        updateActionModal({});
        windowEmitter.emit('station-refresh');
      }
    });
  };

  const cancelAction = () => {
    updateActionModal({});
  };

  const handleSaveInput = () => {
    if (sensorDetails.name === '') {
      updateSuccessMessage('Name must be something');
      return;
    }

    const body = {
      id: sensor.id,
      name: sensorDetails.name,
      description: sensorDetails.description === '' ? null : sensorDetails.description,
      coefficients: sensorDetails.Coefs === '' ? null : sensorDetails.coefficients,
      pollRate: sensorDetails.pollRate === '' ? null : sensorDetails.pollRate,
      unit: sensorDetails.unit === '' ? null : sensorDetails.unit,
    };
    Api.updateSensor(body).then(res => {
      updateSuccessMessage(res.message || res.error);
      setTimeout(() => {
        updateSuccessMessage('');
      }, 15000);
    });
  };

  const handleInputChange = event => {
    setSensorDetails({ ...sensorDetails, [event.target.name]: event.target.value });
  };
  let datestring = 'No readings';
  if (sensor.lastReading) {
    const date = new Date(sensor.lastReading.timestamp);

    datestring = `${date.getDate()}-${
      date.getMonth() + 1
    }-${date.getFullYear()} ${date.getHours()}:${date.getMinutes()}`;
  }

  return (
    <div className={className}>
      {actionModal.id && (
        <AddAction
          action={actionModal}
          actionConditions={actionConditions}
          message={actionMessage}
          save={saveAction}
          cancel={cancelAction}
        />
      )}
      <div className="title">Sensor Details</div>
      <div className="grid">
        <div className="item">
          Station:
          <span>{sensor.stationName}</span>
        </div>
        <div className="item">
          Sensor:
          <span>{sensor.name}</span>
        </div>
        <div className="item">
          Last Reading Time:
          <span>{datestring}</span>
        </div>
        <div className="item">
          Last Reading Value:
          <span>{sensor.lastReading ? sensor.lastReading.value : 'null'}</span>
        </div>

        <div className="item">
          Name:
          <input value={sensorDetails.name} name="name" onChange={handleInputChange} />
        </div>
        <div className="item">
          Description
          <input
            value={sensorDetails.description}
            name="description"
            onChange={handleInputChange}
          />
        </div>
        <div className="item">
          Coefficients:
          <input
            value={sensorDetails.coefficients}
            name="coefficients"
            onChange={handleInputChange}
          />
        </div>
        <div className="item">
          Unit:
          <input value={sensorDetails.unit} name="unit" onChange={handleInputChange} />
        </div>
        <div className="item">
          Poll rate:
          <input value={sensorDetails.pollRate} name="pollRate" onChange={handleInputChange} />
        </div>

        <div className="item">
          <button type="button" onClick={handleSaveInput}>
            save
          </button>
          {successMessage !== '' && <span>{successMessage}</span>}
        </div>
      </div>

      <div className="title">Actions</div>

      <div className="actions-grid">
        <div className="action-row">
          {actionProperties.map(prop => (
            <div className="actions-item" key={prop}>
              {prop}
            </div>
          ))}
          <div />
          <div />
        </div>

        {sensor.actions.map(action => (
          <div className="action-row" key={action.id}>
            <div className="actions-item">{action.value}</div>
            <div className="actions-item">{action.valueType}</div>
            <div className="actions-item">{action.condition}</div>
            <div className="actions-item">{action.action}</div>
            <div className="actions-item">{action.interval}</div>
            <div className="actions-item">{action.metaData}</div>
            <button type="button" onClick={() => editAction(action)}>
              Edit
            </button>
            <button type="button" onClick={() => deleteAction(action)}>
              Delete
            </button>
          </div>
        ))}
        <button className="add-button" type="button" onClick={addAction}>
          +
        </button>
      </div>
    </div>
  );
};

SensorDetails.propTypes = {
  className: PropTypes.string.isRequired,
  sensor: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    coefficients: PropTypes.string,
    health: PropTypes.oneOf(['healthy', 'unhealthy']),
    stationId: PropTypes.number.isRequired,
    stationName: PropTypes.string.isRequired,
    pollRate: PropTypes.number,
    unit: PropTypes.string,
    lastReading: PropTypes.shape({
      timestamp: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    }),
    actions: PropTypes.arrayOf(
      PropTypes.shape({
        condition: PropTypes.string.isRequired,
        interval: PropTypes.string.isRequired,
        action: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
      }),
    ),
  }).isRequired,
};

const styledSensorDetails = styled(SensorDetails)`
  width: 100%;
  padding: 0 3rem;

  .grid {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr 1fr;

    .item {
      height: 40px;

      span {
        padding-left: 5px;
        font-weight: 900;
      }

      input {
        margin-left: 5px;
      }
    }
  }

  .title {
    font-size: 24px;
  }

  .actions-grid {
    display: grid;
    margin-bottom: 36px;

    .action-row {
      display: grid;
      grid-template-columns: repeat(8, 1fr);
      height: 40px;

      &:first-of-type {
        background: #c8ffc8;
      }
    }

    .add-button {
      width: 24px;
      margin-bottom: 10rem;
    }
  }
`;

export default styledSensorDetails;
