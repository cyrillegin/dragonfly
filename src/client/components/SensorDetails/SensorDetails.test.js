import { mount } from 'enzyme';
import React from 'react';
import { act } from '@testing-library/react';
import SensorDetails from './SensorDetails';

jest.useFakeTimers();

jest.mock('../../Api', () => ({
  updateSensor: () => new Promise(res => res([])),
  deleteAction: () => new Promise(res => res()),
}));

describe('SensorDetails', () => {
  let date;

  beforeEach(() => {
    date = global.Date;
    global.Date = class {
      getDate = () => '1';

      getMonth = () => '1';

      getFullYear = () => '1';

      getHours = () => '1';

      getMinutes = () => '1';

      static now() {
        return '1';
      }
    };
  });

  afterEach(() => {
    global.Date = date;
  });

  it('should render a snap shot', async () => {
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <SensorDetails
          sensor={{
            id: 1,
            name: 'test',
            health: 'healthy',
            stationId: 1,
            stationName: 'test',
            lastReading: {
              timestamp: '2020-10-04T20:59:30.176Z',
              value: 1,
            },
            actions: [{ condition: 'test', interval: 'test', action: 'test', id: 1 }],
          }}
        />,
      );
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should test functionality', async () => {
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <SensorDetails
          sensor={{
            id: 1,
            name: 'test',
            health: 'healthy',
            stationId: 1,
            stationName: 'test',
            lastReading: {
              timestamp: '2020-10-04T20:59:30.176Z',
              value: 1,
            },
            actions: [{ condition: 'test', interval: 'test', action: 'test', id: 1 }],
          }}
        />,
      );
      wrapper.find('button').forEach(button => {
        button.simulate('click');
        jest.runAllTimers();
      });
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
