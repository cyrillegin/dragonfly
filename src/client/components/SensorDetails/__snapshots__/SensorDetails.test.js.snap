// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SensorDetails should render a snap shot 1`] = `ReactWrapper {}`;

exports[`SensorDetails should test functionality 1`] = `"<div class=\\"css-66tsks\\"><div class=\\"css-np5om9\\" role=\\"button\\" tabindex=\\"0\\"><div class=\\"modal\\" role=\\"button\\" tabindex=\\"0\\"><div class=\\"title\\">Add Action</div><div class=\\"body\\"><div><span>Value:</span><input name=\\"value\\" value=\\"0\\"></div><div><span>Value Type:</span><select name=\\"valueType\\"><option value=\\"value\\">value</option><option value=\\"timestamp\\">timestamp</option><option value=\\"time\\">time</option><option value=\\"interval\\">interval</option></select></div><div><span> Condition:</span><select name=\\"condition\\"><option value=\\"gt\\">gt</option><option value=\\"gte\\">gte</option><option value=\\"eq\\">eq</option><option value=\\"lt\\">lt</option><option value=\\"lte\\">lte</option></select></div><div><span> Action:</span><select name=\\"action\\"><option value=\\"slack\\">slack</option><option value=\\"wemo\\">wemo</option><option value=\\"camera\\">camera</option></select></div><div><span> Interval:</span><input name=\\"interval\\" type=\\"text\\" value=\\"test\\"></div><div><span>Metadata:</span><input name=\\"metaData\\" type=\\"text\\" value=\\"\\"></div></div><div class=\\"error-message\\"></div><div class=\\"footer\\"><button type=\\"button\\">Save</button><button type=\\"button\\">Cancel</button></div></div></div><div class=\\"title\\">Sensor Details</div><div class=\\"grid\\"><div class=\\"item\\">Station:<span>test</span></div><div class=\\"item\\">Sensor:<span>test</span></div><div class=\\"item\\">Last Reading Time:<span>1-11-1 1:1</span></div><div class=\\"item\\">Last Reading Value:<span>1</span></div><div class=\\"item\\">Name:<input name=\\"name\\" value=\\"test\\"></div><div class=\\"item\\">Description<input name=\\"description\\" value=\\"\\"></div><div class=\\"item\\">Coefficients:<input name=\\"coefficients\\" value=\\"\\"></div><div class=\\"item\\">Unit:<input name=\\"unit\\" value=\\"\\"></div><div class=\\"item\\">Poll rate:<input name=\\"pollRate\\" value=\\"\\"></div><div class=\\"item\\"><button type=\\"button\\">save</button><span></span></div></div><div class=\\"title\\">Actions</div><div class=\\"actions-grid\\"><div class=\\"action-row\\"><div class=\\"actions-item\\">value</div><div class=\\"actions-item\\">value type</div><div class=\\"actions-item\\">condition</div><div class=\\"actions-item\\">action</div><div class=\\"actions-item\\">interval</div><div class=\\"actions-item\\">metaData</div><div></div><div></div></div><div class=\\"action-row\\"><div class=\\"actions-item\\"></div><div class=\\"actions-item\\"></div><div class=\\"actions-item\\">test</div><div class=\\"actions-item\\">test</div><div class=\\"actions-item\\">test</div><div class=\\"actions-item\\"></div><button type=\\"button\\">Edit</button><button type=\\"button\\">Delete</button></div><button class=\\"add-button\\" type=\\"button\\">+</button></div></div>"`;
